### PR TITLE
Add NSGA-II based multi-objective sampler.

### PR DESCRIPTION
**Pull request by sile**
This PR adds a multi-objective sampler based on the NSGA-II algorithm and make it the default multi-objective sampler.
Note that the current implementation is very naive. It could be improved a lot in the future.

# TODO

- [x] Add documentation
- [x] Add tests
- [x] Run a benchmark

### CODE REVIEW

**Comment by sile** [comment]
## Benchmark Result

### NASBench



### ZDT1



### ZDT2



### ZDT3



### ZDT4



### ZDT5



### ZDT6

File: optuna/multi_objective/samplers/_nsga2.py [hunk 1]
```diff
@@ -0,0 +1,254 @@
...
+                    population[i] = population.pop()
+                non_dominated_population.append(individual)
+            else:
+                i += 1
+
+        for x in non_dominated_population:
+            for y in dominates_list[x.number]:
+                dominated_count[y] -= 1
+
+        assert non_dominated_population != []
```
Comment (ytsmiling) [inline optuna/multi_objective/samplers/_nsga2.py:220]:
```suggestion
        assert non_dominated_population
```

File: optuna/multi_objective/samplers/_nsga2.py [hunk 2]
```diff
@@ -0,0 +1,254 @@
...
+        distances[population[0].number] = float("inf")
+        distances[population[-1].number] = float("inf")
+
+        for j in range(1, len(population) - 1):
+            v_high = population[j + 1].values[i]
+            v_low = population[j - 1].values[i]
+            assert v_high is not None
+            assert v_low is not None
+
+            distances[population[j].number] += (v_high - v_low) / width
```
Comment (ytsmiling) [inline optuna/multi_objective/samplers/_nsga2.py:251]: Please rename `distances` to something like `l1_distances` or `manhattan_distances`, or maybe `normalized_l1_distances`.

File: optuna/multi_objective/samplers/_nsga2.py [hunk 3]
```diff
@@ -0,0 +1,254 @@
...
+    ) -> Any:
+        if _PARENTS_KEY not in trial.system_attrs:
+            return self._random_sampler.sample_independent(
+                study, trial, param_name, param_distribution
+            )
+
+        trials = study.get_trials(deepcopy=False)
+        p0, p1 = trial.system_attrs[_PARENTS_KEY]
+
+        param = trials[p0].params.get(param_name, None)
```
[the remark this answers is not in the document]
Comment (sile) [reply optuna/multi_objective/samplers/_nsga2.py:124]: > I think trials are not necessarily sorted by the trial number. 

The API doc of [Study.get_trials()](https://optuna.readthedocs.io/en/stable/reference/study.html#optuna.study.Study.get_trials) says "The returned trials are ordered by trial number", so I think we can rely on the order (however it seems that we should add the restriction to the storage spec too).

> I suggest using trial_id instead of trial.number and access the trial via get_trial method.
> It's also substantially faster than getting whole trials in most DBs excluding in-memory db.

Thanks. It sounds nice.
I wrote the current code because the current Optuna coding convention recommends accessing trials via `Study` rather than directly using the storage interface as much as possible  (it's not well documented though...). However, your suggestion seems sufficiently reasonable so I'll adopt it.

[the remark this answers is not in the document]
Comment (sile) [reply optuna/multi_objective/samplers/_nsga2.py]: Applied the suggestion: https://github.com/optuna/optuna/pull/1163/commits/cca50c1655af4e92f59901af3a63d98c6d0d9300

File: optuna/multi_objective/samplers/_nsga2.py [hunk 4]
```diff
@@ -0,0 +1,254 @@
...
+        if param is None or self._rng.rand() < self._crossover_prob:
+            param = trials[p1].params.get(param_name, None)
```
[the remark this answers is not in the document]
Comment (sile) [reply optuna/multi_objective/samplers/_nsga2.py]: Ditto: https://github.com/optuna/optuna/pull/1163/commits/cca50c1655af4e92f59901af3a63d98c6d0d9300

File: optuna/multi_objective/samplers/_nsga2.py [hunk 5]
```diff
@@ -0,0 +1,254 @@
...
+    def _collect_parent_population(
+        self, study: "multi_objective.study.MultiObjectiveStudy"
+    ) -> Tuple[int, List["multi_objective.trial.FrozenMultiObjectiveTrial"]]:
+        # TODO(ohta): Optimize this method.
+
+        trials = study.get_trials(deepcopy=False)
+        parent_population = []  # type: List[multi_objective.trial.FrozenMultiObjectiveTrial]
+        parent_generation = -1
+        for generation in itertools.count():
+            population = [
```
[the remark this answers is not in the document]
Comment (sile) [reply optuna/multi_objective/samplers/_nsga2.py:147]: Thanks. Updated: https://github.com/optuna/optuna/pull/1163/commits/2a52ce843b00ac268474d61b20c6ed4941848b20

File: optuna/multi_objective/samplers/_nsga2.py [hunk 6]
```diff
@@ -0,0 +1,254 @@
...
+                t
+                for t in trials
+                if (
+                    t.system_attrs.get(_GENERATION_KEY, 0) == generation
+                    and t.state == optuna.trial.TrialState.COMPLETE
+                )
+            ]
+            if len(population) < self._population_size:
+                break
```
[the remark this answers is not in the document]
Comment (sile) [reply optuna/multi_objective/samplers/_nsga2.py:154]: Left the comment to the source code (https://github.com/optuna/optuna/pull/1163/commits/0e0bca869a98d2a38fd2e31fed574e0de23d7c54) as it would be beneficial for future readers.

File: optuna/multi_objective/samplers/_nsga2.py [hunk 7]
```diff
@@ -0,0 +1,254 @@
...
+
+        return parent_generation, parent_population
+
+    def _select_parent(
+        self,
+        study: "multi_objective.study.MultiObjectiveStudy",
+        population: List["multi_objective.trial.FrozenMultiObjectiveTrial"],
+    ) -> "multi_objective.trial.FrozenMultiObjectiveTrial":
+        candidate0 = self._rng.choice(population)
+        candidate1 = self._rng.choice(population)
```
[the remark this answers is not in the document]
Comment (sile) [reply optuna/multi_objective/samplers/_nsga2.py:179]: Added a TODO comment: https://github.com/optuna/optuna/pull/1163/commits/626b7ed2d19ec7b7fee6451491566c36213c7ad7

File: optuna/multi_objective/samplers/_nsga2.py (hunk 2 above)
Comment (sile) [reply optuna/multi_objective/samplers/_nsga2.py]: Thank you for the suggestion. I chose  `manhatten_distances` because it is the most frequently used word in the GitHub repositories among the three names.
Commit: https://github.com/optuna/optuna/pull/1163/commits/cc48c99147b9672e32e98251c206aa0b7664634c

File: tests/multi_objective/samplers_tests/test_nsga2.py [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+        value=None,
+        datetime_start=None,
+        datetime_complete=None,
+        params={},
+        distributions={},
+        user_attrs={},
+        system_attrs={},
+        intermediate_values={i: v for i, v in enumerate(values)},
+    )
+    return multi_objective.trial.FrozenMultiObjectiveTrial(len(values), trial)
```
[the remark this answers is not in the document]
Comment (sile) [reply tests/multi_objective/samplers_tests/test_nsga2.py:161]: Thanks. Added a TODO comment: https://github.com/optuna/optuna/pull/1163/commits/3c9514e9c8d91eae20a27bd078e1f1a13ca87ad1

File: tests/multi_objective/test_trial.py [hunk 9]
```diff
@@ -84,7 +84,10 @@ def objective(trial: optuna.multi_objective.trial.MultiObjectiveTrial) -> List[f
 
 
 def test_system_attrs() -> None:
-    study = optuna.multi_objective.create_study(["maximize", "minimize", "maximize"])
+    study = optuna.multi_objective.create_study(
+        ["maximize", "minimize", "maximize"],
+        sampler=optuna.multi_objective.samplers.RandomMultiObjectiveSampler(),
+    )
```
[the remark this answers is not in the document]
Comment (sile) [reply tests/multi_objective/test_trial.py]: Nice question. I left a comment that explains the reason: https://github.com/optuna/optuna/pull/1163/commits/ed9383ae2cc8202136d2d065aff0db55f615d7c9

File: optuna/multi_objective/samplers/_nsga2.py [hunk 10]
```diff
@@ -0,0 +1,256 @@
...
+        self._rng = np.random.RandomState(seed)
+
+    def infer_relative_search_space(
+        self,
+        study: "multi_objective.study.MultiObjectiveStudy",
+        trial: "multi_objective.trial.FrozenMultiObjectiveTrial",
+    ) -> Dict[str, BaseDistribution]:
+        return {}
+
+    def sample_relative(
```
[the remark this answers is not in the document]
Comment (hvy) [reply optuna/multi_objective/samplers/_nsga2.py:85]: Ah, it's of course because the parents are only determined once per trial... I always get confused by these methods(-names).

File: optuna/multi_objective/samplers/_nsga2.py [hunk 11]
```diff
@@ -0,0 +1,256 @@
...
+    ) -> Dict[str, Any]:
+        parent_generation, population = self._collect_parent_population(study)
+        trial_id = trial._trial_id
+
+        if len(population) == 0:
+            generation = 0
+            study._storage.set_trial_system_attr(trial_id, _GENERATION_KEY, generation)
+        else:
+            p0 = self._select_parent(study, population)
+            p1 = self._select_parent(study, population)
```
Comment (hvy) [inline optuna/multi_objective/samplers/_nsga2.py:99]: For smaller populations, it might be somewhat likely that the same parent is chosen for `p0` and `p1`. How about resampling until they're unique?

Comment (toshihikoyanase) [reply optuna/multi_objective/samplers/_nsga2.py:99]: > How about resampling until they're unique?

IMHO, resampling can frequently occur if the problem is simple and the Pareto-front is narrow.
How about removing `p0` from `population` to sample `p1`?

File: optuna/multi_objective/samplers/_nsga2.py [hunk 12]
```diff
@@ -0,0 +1,256 @@
...
+            Seed for random number generator.
+
+        population_size:
+            Number of individuals (trials) in a generation.
+
+        mutation_prob:
+            Probability of mutating each parameter when creating a new individual.
+
+        crossover_prob:
+            Probability of swapping each parameter of the parents when creating a new individual.
```
[the remark this answers is not in the document]
Comment (toshihikoyanase) [reply optuna/multi_objective/samplers/_nsga2.py:47]: I agree with @hvy.
I thought this probability was to determine whether two chromosomes are recombined or not.

> (which I'd suspect usually is 0.5, equal probability for both parents which also makes me wonder whether this arguments is really needed since the parents are chosen randomly uniformly).

I agree. [This article](https://pdfs.semanticscholar.org/8ed1/d9f5a017c61abfec4089aba828b398aa3e7e.pdf) also explains that this probability is usually 0.5 (see `swapping_probability` in the description of uniform crossover.)

I personally prefer adding `crossover_prob` in order to specify the probability of whether a crossover should occur, and fixing the `swapping_probability` to `0.5`. In this case, the value of `crossover_prob` seems to be 0.9 in Deb's paper.

File: tests/multi_objective/samplers_tests/test_nsga2.py [hunk 13]
```diff
@@ -0,0 +1,161 @@
+from collections import defaultdict
+import pytest
```
[the remark this answers is not in the document]
Comment (sile) [reply tests/multi_objective/samplers_tests/test_nsga2.py]: Nice catch. Fixed: https://github.com/optuna/optuna/pull/1163/commits/817c67e6f8f02f5c1a13a70a32bf60285e5cb448

File: tests/multi_objective/samplers_tests/test_nsga2.py [hunk 14]
```diff
@@ -0,0 +1,161 @@
...
+
+
+def test_population_size() -> None:
+    # Set `population_size` to 10.
+    sampler = multi_objective.samplers.NSGAIIMultiObjectiveSampler(population_size=10)
+
+    study = multi_objective.create_study(["minimize"], sampler=sampler)
+    study.optimize(lambda t: [t.suggest_uniform("x", 0, 9)], n_trials=40)
+
+    generations = defaultdict(int)
```
[the remark this answers is not in the document]
Comment (sile) [reply tests/multi_objective/samplers_tests/test_nsga2.py]: Thank you for the information. I did not know `Counter`.
I replaced `defaultdict` with `Counter`  and the new code seems a bit simpler than before: https://github.com/optuna/optuna/pull/1163/commits/f81184a1a925ec6afaa7388b306398252f7fa45f

File: optuna/multi_objective/samplers/_nsga2.py (hunk 11 above)
Comment (sile) [reply optuna/multi_objective/samplers/_nsga2.py]: Thank you for your advice! I adopted the logic suggested by @toshihikoyanase at the commit https://github.com/optuna/optuna/pull/1163/commits/eac981f1092e324bfd9beb60638368b17d27ac3f. 
Besides https://github.com/optuna/optuna/pull/1163#discussion_r424243213 and https://github.com/optuna/optuna/pull/1163#discussion_r424242441 were also addressed in the same commit.

File: optuna/multi_objective/samplers/_nsga2.py [hunk 15]
```diff
@@ -0,0 +1,256 @@
...
+            return self._random_sampler.sample_independent(
+                study, trial, param_name, param_distribution
+            )
+
+        return param
+
+    def _collect_parent_population(
+        self, study: "multi_objective.study.MultiObjectiveStudy"
+    ) -> Tuple[int, List["multi_objective.trial.FrozenMultiObjectiveTrial"]]:
+        # TODO(ohta): Optimize this method.
```
[the remark this answers is not in the document]
Comment (sile) [reply optuna/multi_objective/samplers/_nsga2.py:137]: > an easy thing you could do is to replace the trials list with a dictionary {"generation": "trials"} not having to go through all trials in each outer loop.

Sounds good. I adopted the approach at the commit https://github.com/optuna/optuna/pull/1163/commits/c0c3eff0a4dfb697cb1c76f55bd33a3d3946ff86.

> Just wondering but do you have any concrete ideas?

I thought to introduce a caching mechanism like shown below:
```diff
diff --git a/optuna/multi_objective/samplers/_nsga2.py b/optuna/multi_objective/samplers/_nsga2.py
index 17176939..ffbc76ff 100644
--- a/optuna/multi_objective/samplers/_nsga2.py
+++ b/optuna/multi_objective/samplers/_nsga2.py
@@ -1,4 +1,5 @@
 from collections import defaultdict
+import hashlib
 import itertools
 from typing import Any
 from typing import DefaultDict
@@ -19,6 +20,7 @@ from optuna.multi_objective.samplers import BaseMultiObjectiveSampler
 # Define key names of `Trial.system_attrs`.
 _GENERATION_KEY = "multi_objective:nsga2:generation"
 _PARENTS_KEY = "multi_objective:nsga2:parents"
+_POPULATION_KEY = "multi_objective:nsga2:population:{}"
 
 
 @experimental("1.5.0")
@@ -134,19 +136,27 @@ class NSGAIIMultiObjectiveSampler(BaseMultiObjectiveSampler):
     def _collect_parent_population(
         self, study: "multi_objective.study.MultiObjectiveStudy"
     ) -> Tuple[int, List["multi_objective.trial.FrozenMultiObjectiveTrial"]]:
-        # TODO(ohta): Optimize this method.
-
+        trials = study.get_trials(deepcopy=False)
+        generation_to_runnings = defaultdict(list)
         generation_to_population = defaultdict(list)
-        for trial in study.get_trials(deepcopy=False):
+        for trial in trials:
+            if _GENERATION_KEY not in trial.system_attrs:
+                continue
+
+            generation = trial.system_attrs[_GENERATION_KEY]
+
             if trial.state != optuna.trial.TrialState.COMPLETE:
+                if trial.state == optuna.trial.TrialState.RUNNING:
+                    generation_to_runnings[generation].append(trial.number)
                 continue
 
-            generation = trial.system_attrs.get(_GENERATION_KEY, 0)
             generation_to_population[generation].append(trial)
 
+        hasher = hashlib.sha256()
         parent_population = []  # type: List[multi_objective.trial.FrozenMultiObjectiveTrial]
         parent_generation = -1
-        for generation in itertools.count():
+        while True:
+            generation = parent_generation + 1
             population = generation_to_population[generation]
 
             # Under multi-worker settings, the population size might become larger than
@@ -154,22 +164,47 @@ class NSGAIIMultiObjectiveSampler(BaseMultiObjectiveSampler):
             if len(population) < self._population_size:
                 break
 
+            for trial in generation_to_runnings[generation]:
+                hasher.update(bytes(str(trial.number), "utf-8"))
+
+            population_cache_key = _POPULATION_KEY.format(hasher.hexdigest())
+            parent_generation, parent_population_numbers = study.system_attrs.get(
+                population_cache_key, (-1, [])
+            )
+            if parent_generation >= generation:
+                parent_population = [trials[n] for n in parent_population_numbers]
+                continue
+
             population.extend(parent_population)
-            parent_population = []
+            parent_population = self._select_elite_population(study, population)
             parent_generation = generation
 
-            population_per_rank = _fast_non_dominated_sort(population, study.directions)
-            for population in population_per_rank:
-                if len(parent_population) + len(population) < self._population_size:
-                    parent_population.extend(population)
-                else:
-                    n = self._population_size - len(parent_population)
-                    _crowding_distance_sort(population)
-                    parent_population.extend(population[:n])
-                    break
+            if len(generation_to_runnings[parent_generation]) == 0:
+                study.set_system_attr(
+                    population_cache_key,
+                    (parent_generation, [t.number for t in parent_population]),
+                )
 
         return parent_generation, parent_population
 
+    def _select_elite_population(
+        self,
+        study: "multi_objective.study.MultiObjectiveStudy",
+        population: List["multi_objective.trial.FrozenMultiObjectiveTrial"],
+    ) -> List["multi_objective.trial.FrozenMultiObjectiveTrial"]:
+        elite_population = []
+        population_per_rank = _fast_non_dominated_sort(population, study.directions)
+        for population in population_per_rank:
+            if len(elite_population) + len(population) < self._population_size:
+                elite_population.extend(population)
+            else:
+                n = self._population_size - len(elite_population)
+                _crowding_distance_sort(population)
+                elite_population.extend(population[:n])
+                break
+
+        return elite_population
+
     def _select_parent(
         self,
         study: "multi_objective.study.MultiObjectiveStudy",
```

I'd like the comment to keep as it is because I'll create a PR to introduce the cache just after this PR is merged.
(To summarize the above logic in a sentence is a bit hard to me...)

File: optuna/multi_objective/samplers/_nsga2.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (sile) [reply optuna/multi_objective/samplers/_nsga2.py]: @hvy @toshihikoyanase Thanks! You're definitely right. I renamed the existing `crossover_prob` with `swapping_prob` and added the correct `crossover_prob` parameter. The default value of each parameter are also fixed (https://github.com/optuna/optuna/pull/1163/commits/fbaf12e03546c47d6527b1574dd3a19c8c2d777d).

[the remark this answers is not in the document]
Comment (sile) [reply optuna/multi_objective/samplers/_nsga2.py]: Besides, I modified the type of `mutation_prob` to allow taking `None` at the commit https://github.com/optuna/optuna/pull/1163/commits/89457170d9942289d1ead5f2600c4f0493e4d6cf. If `None` is passed,  `1 / number of parent trial parameters` is used as the value. This behavior seems to align to  Deb's paper.

File: optuna/multi_objective/samplers/_nsga2.py [hunk 16]
```diff
@@ -0,0 +1,278 @@
...
+        p0_id, p1_id = trial.system_attrs[_PARENTS_KEY]
+        p0 = study._storage.get_trial(p0_id)
+        p1 = study._storage.get_trial(p1_id)
+
+        param = p0.params.get(param_name, None)
+        if param is None or self._rng.rand() < self._swapping_prob:
+            param = p1.params.get(param_name, None)
+
+        mutation_prob = (
+            self._mutation_prob if self._mutation_prob is not None else 1.0 / len(p0.params)
```
Comment (hvy) [inline optuna/multi_objective/samplers/_nsga2.py:146]: An edge case but this could result in a zero division error. How about resorting to `p1.params` or making sure it's always mutated in such cases?

File: tests/multi_objective/test_trial.py [hunk 17]
```diff
@@ -84,7 +84,13 @@ def objective(trial: optuna.multi_objective.trial.MultiObjectiveTrial) -> List[f
 
 
 def test_system_attrs() -> None:
-    study = optuna.multi_objective.create_study(["maximize", "minimize", "maximize"])
+    # We use `RandomMultiObjectiveSampler` here because the default `NSGAIimultiobjectivesampler`
+    # sampler sets its own system attributes when sampling (these attributes would become noise in
+    # this test case).
```
[the remark this answers is not in the document]
Comment (sile) [reply tests/multi_objective/test_trial.py:89]: Thanks!

File: optuna/multi_objective/samplers/_nsga2.py [hunk 18]
```diff
@@ -0,0 +1,278 @@
...
+    """
+
+    def __init__(
+        self,
+        seed: Optional[int] = None,
+        population_size: int = 50,
+        mutation_prob: Optional[float] = None,
+        crossover_prob: float = 0.9,
+        swapping_prob: float = 0.5,
+    ) -> None:
```
[the remark this answers is not in the document]
Comment (sile) [reply optuna/multi_objective/samplers/_nsga2.py:64]: That makes sense. Fixed: https://github.com/optuna/optuna/pull/1163/commits/be0672c576d19f82cebdc8ad825dc37ed4bf3a19

File: optuna/multi_objective/samplers/_nsga2.py (hunk 16 above)
Comment (sile) [reply optuna/multi_objective/samplers/_nsga2.py]: Nice catch! Fixed: https://github.com/optuna/optuna/pull/1163/commits/ed3475f6f47ff86d3c0f40d956827f9ecc023ef6

File: optuna/multi_objective/samplers/_nsga2.py [hunk 19]
```diff
@@ -0,0 +1,281 @@
...
+    which is a well known, fast and elitist multi-objective genetic algorithm.
+
+    For further information about NSGA-II, please refer to the following paper:
+
+    - `A fast and elitist multiobjective genetic algorithm: NSGA-II
+      <https://ieeexplore.ieee.org/document/996017>`_
+
+    Args:
+        seed:
+            Seed for random number generator.
```
Comment (hvy) [inline optuna/multi_objective/samplers/_nsga2.py:38]: Let's align the argument order putting `seed` last.

File: optuna/multi_objective/samplers/_nsga2.py [hunk 20]
```diff
@@ -0,0 +1,281 @@
...
+
+        population_size:
+            Number of individuals (trials) in a generation.
+
+        mutation_prob:
+            Probability of mutating each parameter when creating a new individual.
+            If :obj:`None` is specified, the value ``1.0 / len(parent_trial.params)`` is used
+            where ``parent_trial`` is the first parent trial of the target individual.
```
Comment (hvy) [inline optuna/multi_objective/samplers/_nsga2.py:46]: How about abstracting it a bit, now that we also try the "second" parent?
```suggestion
            where ``parent_trial`` is the parent trial of the target individual.
```

**Comment by sile** [comment]
> I'm wondering what you think about running a second experiment to compare the results with the previous logic with 1.0 crossover probability with biased swapping.

@hvy It hasn't been uploaded, but I actually have run the benchmark (the result is shown below).
It seems difficult to say which is better from the result, though.

### NASBench



### ZDT1



### ZDT2



### ZDT3



### ZDT4



### ZDT5



### ZDT6

File: optuna/multi_objective/samplers/_nsga2.py (hunk 19 above)
Comment (sile) [reply optuna/multi_objective/samplers/_nsga2.py]: Oops, fixed: https://github.com/optuna/optuna/pull/1163/commits/012f02db2acc5299caf43f2633e4abbc1f3f7e9f

**Comment by sile** [comment]
@hvy Thank you for your careful review...! I reflected your comments, so PTAL 🙇
